### PR TITLE
[GOBBLIN-2090] delete deadline triggers in all hosts

### PR DESCRIPTION
**Pull request by arjun4084346**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-2090

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
i) It changes DagManagementDagActionStoreChangeMonitor and delete triggers whenever deadline enforcement dag actions are deleted.
ii) Removed EnforceJobStartDeadlineDagAction when a job/dag is cancelled, so that the job that never started (and hence able to clear EnforceJobStartDeadlineDagAction), can clean it now.
iii) removed unused enums CANCEL and RETRY in DagActionStore
iv) also fixes several unit tests and did refactoring
v) fixed codestyle changes suggested by intelliJ whatever came across during the work

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
DagManagementDagActionStoreChangeMonitorTest

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-modules/gobblin-kafka-09/src/test/java/org/apache/gobblin/runtime/DagManagementDagActionStoreChangeMonitorTest.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+  }
+
+  @BeforeClass
+  public void setUp() throws Exception {
+    doNothing().when(dagActionReminderScheduler).unscheduleReminderJob(any());
+
+  }
+
+  /**
+   * Tests process message with a DELETE type message which should be ignored regardless of the flow information.
```
Comment (umustafi) [inline gobblin-modules/gobblin-kafka-09/src/test/java/org/apache/gobblin/runtime/DagManagementDagActionStoreChangeMonitorTest.java:106]: is this desc accurate? the delete is not ignored anymore

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/DagActionStoreChangeMonitor.java [hunk 2]
```diff
@@ -224,31 +224,41 @@ protected void processMessage(DecodeableKafkaRecord message) {
...
-            flowExecutionId, dagActionType);
-        this.unexpectedErrors.mark();
-      } else if (operation.equals("DELETE")) {
-        log.debug("Deleted dagAction from DagActionStore: {}", dagAction);
-      } else {
-        log.warn("Received unsupported change type of operation {}. Expected values to be in [INSERT, UPDATE, DELETE]",
-            operation);
-        this.unexpectedErrors.mark();
-        return;
+      switch (operation) {
```
Comment (umustafi) [inline gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/DagActionStoreChangeMonitor.java:237]: nice improvement

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/DagManagementDagActionStoreChangeMonitor.java [hunk 3]
```diff
@@ -37,16 +40,56 @@
...
   private final DagManagement dagManagement;
+  @VisibleForTesting @Getter
+  private final DagActionReminderScheduler dagActionReminderScheduler;
 
   // Note that the topic is an empty string (rather than null to avoid NPE) because this monitor relies on the consumer
   // client itself to determine all Kafka related information dynamically rather than through the config.
   public DagManagementDagActionStoreChangeMonitor(Config config, int numThreads,
       FlowCatalog flowCatalog, Orchestrator orchestrator, DagManagementStateStore dagManagementStateStore,
-      boolean isMultiActiveSchedulerEnabled, DagManagement dagManagement) {
+      boolean isMultiActiveSchedulerEnabled, DagManagement dagManagement, DagActionReminderScheduler dagActionReminderScheduler) {
```
Comment (umustafi) [inline gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/DagManagementDagActionStoreChangeMonitor.java:50]: do u need to make sure initialization order in guide works here?

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/DagManagementDagActionStoreChangeMonitor.java [hunk 4]
```diff
@@ -37,16 +40,56 @@
...
+    // We only expect INSERT and DELETE operations done to this table. INSERTs correspond to any type of
+    // {@link DagActionStore.FlowActionType} flow requests that have to be processed.
+    try {
+      switch (operation) {
+        case "INSERT":
+          handleDagAction(dagAction, false);
+          break;
+        case "UPDATE":
+          // TODO: change this warning message and process updates if for launch or reevaluate type
+          log.warn("Received an UPDATE action to the DagActionStore when values in this store are never supposed to be "
```
Comment (umustafi) [inline gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/DagManagementDagActionStoreChangeMonitor.java:70]: this is not true any longer like ur saying so we should update this

File: gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/proc/LaunchDagProcTest.java [hunk 5]
```diff
@@ -180,4 +175,29 @@ public static Dag<JobExecutionPlan> buildDagWithMultipleNodesAtDifferentLevels(S
...
   }
+
+  public static void mockDMSSCommonBehavior(DagManagementStateStore dagManagementStateStore) throws IOException, SpecNotFoundException {
+    doReturn(FlowSpec.builder().build()).when(dagManagementStateStore).getFlowSpec(any());
+    doNothing().when(dagManagementStateStore).tryAcquireQuota(any());
+    doReturn(true).when(dagManagementStateStore).releaseQuota(any());
+  }
+
+  public static TopologySpec buildNaiveTopologySpec(String specUriInString) {
+    String specStoreDir = "/tmp/specStoreDir";
```
Comment (umustafi) [inline gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/proc/LaunchDagProcTest.java:186]: add comments

File: gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/proc/ResumeDagProcTest.java [hunk 6]
```diff
@@ -93,25 +81,26 @@ public void resumeDag()
...
-      } catch (ExecutionException | InterruptedException e) {
-        throw new RuntimeException(e);
-      }
-    }).filter(sp -> specProducer != sp).collect(Collectors.toList());
     int expectedNumOfResumedJobs = 1; // = number of resumed nodes
 
-    Mockito.verify(specProducer, Mockito.times(expectedNumOfResumedJobs)).addSpec(any());
+    // only the current job should have run
+    // we cannot check the spec producers if addSpec is called on them, because in resumeDagProc, a dag is stored in DMSS
+    // and retrieved, hence it goes through serialization/deserialization; in this process  the SpecProducer objects in
```
Comment (umustafi) [inline gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/proc/ResumeDagProcTest.java:97]: nit: extra space before "the specProducer"

Comment (umustafi) [reply gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/proc/ResumeDagProcTest.java:97]: also change to use multi line comment

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java [hunk 7]
```diff
@@ -228,12 +225,12 @@ protected void processMessage(DecodeableKafkaRecord<byte[],byte[]> message) {
             this.eventProducer.emitObservabilityEvent(jobStatus);
           }
 
-          if (this.dagProcEngineEnabled) {
+          if (this.dagProcEngineEnabled && isJobLevelStatus(jobName)) {
```
Comment (umustafi) [inline gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java:228]: do u have test for isJobLevelStatus or this monitoring method?

File: gobblin-modules/gobblin-kafka-09/src/test/java/org/apache/gobblin/runtime/DagManagementDagActionStoreChangeMonitorTest.java (hunk 1 above)
Comment (arjun4084346) [reply gobblin-modules/gobblin-kafka-09/src/test/java/org/apache/gobblin/runtime/DagManagementDagActionStoreChangeMonitorTest.java:106]: bad copy paste

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/DagManagementDagActionStoreChangeMonitor.java (hunk 3 above)
Comment (arjun4084346) [reply gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/DagManagementDagActionStoreChangeMonitor.java:50]: DagActionReminderScheduler only needs StdSchedulerFactory, which does not need any other dependency for initialization, so there should not be any circular dependency.
If creation of DagActionReminderScheduler were indirectly dependent on DagManagementDagActionStoreChangeMonitorFactory, that would have been a problem.

File: gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java (hunk 7 above)
Comment (arjun4084346) [reply gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/KafkaJobStatusMonitor.java]: added in KafkaAvroJobStatusMonitorTest

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/proc/DagProcUtils.java [hunk 8]
```diff
@@ -210,4 +211,36 @@ public static long getDefaultJobStartDeadline(Config config) {
...
+
+  public static boolean isJobLevelStatus(String jobName) {
+    return !jobName.equals(JobStatusRetriever.NA_KEY);
+  }
+
+  public static void removeEnforceJobStartDeadlineDagAction(DagManagementStateStore dagManagementStateStore, String flowGroup,
+      String flowName, long flowExecutionId, String jobName) {
+    DagActionStore.DagAction enforceJobStartDeadlineDagAction = new DagActionStore.DagAction(flowGroup, flowName,
+        flowExecutionId, jobName, DagActionStore.DagActionType.ENFORCE_JOB_START_DEADLINE);
+    log.info("Deleting {}", enforceJobStartDeadlineDagAction);
```
Comment (umustafi) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/proc/DagProcUtils.java:223]: "Deleting dag action {}" (easier to find)

**Review by umustafi — APPROVED** [review]

**Pull request merged**